hasher-matcher-actioner/hmalib/common/content_models.py

```diff
@@ -0,0 +1,222 @@
+            ProjectionExpression=cls.DEFAULT_PROJ_EXP,
+        ).get("Items", [])
+
+        return cls._result_item_to_action_record(items)
+
+    @classmethod
+    def _result_item_to_action_record(
+        cls,
+        items: t.List[t.Dict],
+    ) -> t.List["ActionEventRecord"]:
+        return [
+            ActionEventRecord(
+                content_id=cls.remove_content_key_prefix(
+                    item["PK"],
+                ),
+                time_performed=datetime.datetime.fromisoformat(
+                    item["SK"][len(cls.ACTION_TIME_PREFIX) :]
+                ),
+                action_label=item["ActionLabel"],
+                action_rules=item["ActionRules"],
+            )
+            for item in items
+        ]
+
+
+@dataclass
+class ContentObject(ContentObjectBase, JSONifiable):
+    """
+    Content object that stores the values related to a specific
+    content_id and content_type.
+    Note/TODO: Something like an override flag should be added to the submit API
+    as ContentRef equality check on bytes is impractical so instead
+    something like:
+     "This content_id exists, [x] overwrite? (I know what you're doing)"
+    is likely better.
+    """
+
+    CONTENT_TYPE_PREFIX = "content_type#"
+
+    content_type: str
+    content_ref: str
+    content_ref_type: str
+    additional_fields: t.Optional[t.Set]
+    submissions: t.List
+    created_on: datetime.datetime
+    updated_at: datetime.datetime
+
+    @staticmethod
+    def get_dynamodb_content_type_key(content_type: str) -> str:
+        return f"{ContentObject.CONTENT_TYPE_PREFIX}{content_type}"
+
+    def to_json(self) -> t.Dict:
+        """
+        TODO docstring
+        """
+        result = asdict(self)
+        result.update(
+            additional_fields=list(
+                self.additional_fields if self.additional_fields else set()
+            ),
+            submissions=[s.isoformat() for s in self.submissions],
+            created_on=self.created_on.isoformat(),
+            updated_at=self.updated_at.isoformat(),
+        )
+        return result
+
+    def write_to_table(self, table: Table):
+        """
+        Write operations for this object need to be special cased (to avoid overwritting)
+        Therefore we do not implement `to_dynamodb_item`
+
+        If you're curious it would ~look like this:
+        def to_dynamodb_item(self) -> dict:
+            return {
+                "PK": self.get_dynamodb_content_key(self.content_id),
+                "SK": self.get_dynamodb_content_type_key(self.content_type),
+                "ContentRef": self.content_ref,
+                "ContentRefType": self.content_ref_type,
+                "AdditionalFields": self.additional_fields,
+                "Submissions": [s.isoformat() for s in self.submissions],
+                "CreatedOn": self.created_on.isoformat(),
+                "UpdatedAt": self.updated_at.isoformat(),
+            }
+        """
+        table.update_item(
```

Comment:
`put_item` also takes ConditionExpressions AFAIK. Perhaps explain why you are using `update_item`?

Reply:
I don't think it has `UpdateExpression` which is needed for the SET and ADD stuff.